src/System.Linq/tests/ToHashSetTests.cs

```diff
@@ -70,5 +70,363 @@ public void ThrowOnNullSource()
         {
             Assert.Throws<ArgumentNullException>(() => ((IEnumerable<object>)null).ToHashSet());
         }
+
+        [Fact]
+        public void WithConcat()
+        {
+            var first = Enumerable.Range(0, 50);
+            var second = Enumerable.Range(20, 50);
+
+            HashSet<int> result = first.Concat(second).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(70, result.Count);
```

Comment:
@TiagoSoczek I am not sure these tests are rigorous enough because they do not actually validate the contents of the `HashSet`, only the fact that it is non-null and that it has a certain count. In addition, they do not validate the HashSet has the correct comparer.

You might find it easier to structure the tests like this instead:

```cs
[Theory]
[MemberData(nameof(WorksAsExpectedData))]
public void WorksAsExpected(IEnumerable<int> source)
{
    Assert.All(TestComparers, comparer =>
    {
        var expected = new HashSet<int>(source, comparer);
        var actual = source.ToHashSet(comparer);
        Assert.Equal(expected, actual); // Implies they have equal Counts too
        Assert.Equal(expected.Comparer, actual.Comparer);
    });
}

public static IEnumerable<object[]> WorksAsExpectedData()
{
    yield return new object[] { Range(0, 50).Concat(Range(20, 50)) };
    yield return new object[] { Range(0, 0).DefaultIfEmpty() };
    ...
}

private static IEnumerable<IEqualityComparer<int>> TestComparers =>
    new[] { null, EqualityComparer<int>.Default, new CustomComparer<int>() };
```

That way it's easier to add new test cases and you fully validate what is returned from the method.